### DISCUSSION TITLE
Displaying git status

### DISCUSSION THREAD

**Discussion by janbuchar** [sayanarijit/xplr#529] [General]
Is there any way to display the git status of the files listed in directories tracked by git? I could imagine something like an M for modified files, U for untracked files, etc.

Hope there already isn'a a plugin that I overlooked :slightly_smiling_face:

**Comment by sayanarijit** [comment]
I don't think there's a plugin yet. The way I use git, never felt the need to integrate it with xplr tightly. For loose integration, it should be very easy to run "git status" or spawn something like gitui or lazygit in the current directory by pressing a key (see awesome hacks in the docs for examples).

**Reply by janbuchar** [reply]
And if I wanted to add the status directly to xplr, would it be tough to script? Where should I start looking if I wanted to try?

**Reply by sayanarijit** [reply]
Probably [use this as a reference](https://github.com/jarun/nnn/blob/31b83ef1f06905aec05008a3881f01e0b3361414/patches/gitstatus/mainline.diff), then look at [on_directory_change](https://xplr.dev/en/configuration#hooks) where you can hook the [CallLuaSilently](https://xplr.dev/en/messages#callluasilently) message to run `git status` on directory change, and collect the git information in a Lua variable (see any lua plugin) [doc](https://xplr.dev/en/lua-function-calls). ~Ideally, it should be done by before_directory_change hook, but that hook isn't available yet (probably coming in next release), so~ you'll need to [refresh](https://xplr.dev/en/messages#refresh) the screen.

See [this doc](https://xplr.dev/en/column-renderer) to learn how to display custom columns.

If the on_directory_change hook with Refresh doesn't work as expected for some reason, for now, you'll need to override the `right` key to hook it [see this plugin](https://github.com/sayanarijit/tri-pane.xplr/blob/main/init.lua).

Feel free to ping in discord/matrix if you need help. People are helpful there.

**Reply by sayanarijit** [reply]
Oh, don't forget to see the [default init.lua](https://github.com/sayanarijit/xplr/blob/main/src/init.lua)

**Reply by sayanarijit** [reply]
Example of a directory visit counter:

```lua

local c = {}

table.insert(xplr.config.general.table.header.cols, 1, { format = "visits" })

table.insert(
  xplr.config.general.table.row.cols,
  1,
  { format = "custom.fmt_directory_visits" }
)

table.insert(xplr.config.general.table.col_widths, 1, { Length = 3 })

xplr.fn.custom.on_directory_change = function(app)
  c[app.pwd] = (c[app.pwd] or 0) + 1
  return {
    "Refresh",
  }
end

xplr.fn.custom.fmt_directory_visits = function(m)
  return tostring(c[m.absolute_path] or 0)
end

return {
  on_directory_change = {
    { CallLuaSilently = "custom.on_directory_change" },
  },
}
```
